ophidian/parser/Lef.h

```diff
@@ -0,0 +1,217 @@
+		};
+
+		std::string name; ///< Name of the site
+		std::string class_; ///< Class of the site
+		double x; ///< Width of the site
+		double y; ///< Height of the site
+
+		/// Stores the site's symmetry in accordance to the site::symmetries enumeration
+		char symmetry {0};
+		void setXsymmetry(); ///< Sets the X symmetry bit
+		void setYsymmetry(); ///< Sets the Y symmetry bit
+		void set90symmetry(); ///< Sets the R90 symmetry bit
+		// void addRowPattern(const char* name, int orient);
+	};
+
+	/**
+	 * A structure to represent a layer
+	 */
+	struct layer {
+		/// An enumeration to represent all the possible directions of a layer
+		enum directions {
+			NOT_ASSIGNED, HORIZONTAL, VERTICAL
+		};
+		std::string name; ///< Name of the layer
+		std::string type; ///< Type of the layer
+		directions direction; ///< Direction of the layer
+		double pitch; ///< Pitch of the layer
+		double width; ///< Width of the layer
+	};
+
+	/**
+	 * A structure to represent a rectangle
+	 */
+	struct rect {
+		double xl; ///< x coordinate of the first point
+		double yl; ///< y coordinate of the first point
+		double xh; ///< x coordinate of the second point
+		double yh; ///< y coordinate of the second point
+	};
+
+	/**
+	 * A structure to represent a port
+	 */
+	struct port {
+		std::vector<std::string> layers; ///< A vector with all the names of all the port layers
+		std::vector<rect> rects; ///< A vector with all the rects of the port
+	};
+
+	/**
+	 * A structure to represent a pin
+	 */
+	struct pin {
+		/// An enumeration to represent all the possible directions of a pin
+		enum directions {
+			INPUT, OUTPUT, INOUT, NA
+		};
+
+		std::string name; ///< The pin name
+		directions direction {NA}; ///< The pin's direction in accordance to pin::directions
+		std::vector<port> ports; ///< A vector with all the pin ports
+	};
+
+	/**
+	 * A structure to represent the size of a macro
+	 */
+	struct macro_size {
+		double x; ///< Width of the macro
+		double y; ///< Height of the macro
+	};
+
+	/**
+	 * A structure to represent the foreign property of a macro
+	 */
+	struct macro_foreign {
+		std::string name; ///< Foreign cell name
+		double x; ///< Offset in the x coordinate
+		double y; ///< Offset in the y coordinate
+	};
+
+	/**
+	 * A structure to represent the macro obstructions
+	 */
+	struct obs {
+		/// Map with all the rectangles of obstruction
+		std::map< std::string, std::vector<rect> > layer2rects;
+	};
+
+	/**
+	 * A structure to represent a macro
+	 */
+	struct macro {
+		std::string name; ///< Name of the macro
+		std::string class_; ///< Class of the macro
+		std::vector<pin> pins; ///< Vector with all the macro pins
+		macro_foreign foreign; ///< Struct with the foreign propertiy
+		macro_size size; ///< Struct with the size
+		std::string site; ///< Site name
+		obs obses; ///< Struct with the macro rectangle geometry
+		macro_size origin; ///< Struct containing the origin property
+	};
+
+private:
+	LefDefParser::lefiUnits m_units;
```

Comment:
`struct Impl;`
`std::unique_ptr<Impl> this_;`

In Lef.cpp:
`struct Lef::Impl {
LefDefParser::lefiUnits units_;
};`
`Lef::Lef() : this_(new Impl){}`
`double Lef::databaseUnits() const {
    return this_->m_units.databaseNumber();
}
`